DOC: add changelog for Shapely 2.0 (#1442)
shapely/shapely: docs/release/2.x.rst

> Releases 2.x

Reviewer: ```suggestion
Version 2.x
```

++++++++++++

.. _version-2-0-0:

Version 2.0.0 (in progress)
---------------------------

Shapely 2.0 version is a major release featuring a complete refactor of the

> internals and new vectorized (element-wise) array operations providing

Reviewer: ```suggestion
internals and new vectorized (element-wise) array operations, providing
```

been refactored to expose GEOS functionality through a Python C extension
module that is compiled in advance.

The pointer to the actual GEOS Geometry object is stored in a lightweight
`Python extension type <https://docs.python.org/3/extending/newtypes_tutorial.html>`__.
A single `Geometry` Python extension type is defined in C wrapping a
`GEOSGeometry` pointer. This extension type is further subclassed in Python
to provide the geometry type-specific classes from Shapely (Point,
LineString, Polygon, etc).
The GEOS pointer is accessible from C as a static attribute of the Python
object (an attribute of the C struct that makes up a Python object), which
enables using vectorized functions within C and thus avoiding Python overhead
while looping over an array of geometries (see next section).

Vectorized (element-wise) geometry operations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Before the 2.0 release, Shapely only provided an interface for scalar
(individual) geometry objects. Users had to loop over individual geometries
within an array of geometries and call scalar methods or properties, which is
both more verbose to use and has a large performance overhead.

> Shapely 2.0 exposes all GEOS operations as vectorized functions that operate

Reviewer: ```suggestion
Shapely 2.0 exposes GEOS operations as vectorized functions that operate
```

("all" implies we're completely caught up to the GEOS C API)

etc) are now available in the top-level namespace as well. Thus it is no
longer needed to import those from the ``shapely.geometry`` submodule.

The following::

  from shapely.geometry import Point

can be replaced with::

  from shapely import Point

or::

  import shapely
  shapely.Point(...)

Note: for backwards compatibility (and being able to write code that works
for both <=1.8 and >2.0), those classes still remain accessible from the
``shapely.geometry`` submodule as well.

More informative repr with truncated WKT
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The repr (``__repr__``) of Geometry objects has been simplified and improved
to include a descriptive Well-Known-Text (WKT) formatting. Instead of showing
the class name and id::

  >>> Point(0, 0)
  <shapely.geometry.point.Point at 0x7f0b711f1310>

we now get::

  >>> Point(0, 0)
  <POINT (0 0)>

For large geometries with many coordinates, the output gets truncated to 80
characters.

Support for fixed precision model for geometries and in overlay functions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

GEOS 3.9.0 overhauled the overlay operations (union, intersection,

> (symmetric) difference): a complete rewrite, dubbed "OverlayNG", provides a

Reviewer: ```suggestion
(symmetric) difference).  A complete rewrite, dubbed "OverlayNG", provides a
```

more robust implementation (no more TopologyExceptions even on valid input),
the ability to specify the output precision model, and significant
performance optimizations. When installing Shapely with GEOS >= 3.9 (which is
the case for PyPI wheels and conda-forge packages), you automatically get

> those improvements already (also for previous versions of Shapely) when using

Reviewer: ```suggestion
these improvements (also for previous versions of Shapely) when using
```

order as the geometries.

In addition, Shapely 2.0 includes several improvements to ``STRtree``:

* Directly include predicate evaluation in :meth:`.STRtree.query` by
  specifying the ``predicate`` keyword. If a predicate is provided, tree
  geometries with bounding boxes that overlap the bounding boxes of the input
  geometries are further filtered to those that meet the predicate (using
  prepared geometries under the hood for efficiency).
* Query multiple input geometries (spatial join style) with
  :meth:`.STRtree.query` by passing an array of geometries. In this case, the
  return value is a 2D array with shape (2, n) where the subarrays correspond
  to the indices of the input geometries and indices of the tree geometries
  associated with each.
* A new :meth:`.STRtree.query_nearest` method was added, returning the index
  of the nearest geometries in the tree for each input geometry. Compared to
  :meth:`.STRtree.nearest`, which only returns the index of a single nearest
  geometry for each input geometry, this new methods allows for:

  * returning all equidistant nearest geometries,
  * excluding nearest geometries that are equal to the input,

>   * specifying an ``max_distance`` to limit the search radius potentially

Reviewer: ```suggestion
  * specifying an ``max_distance`` to limit the search radius, potentially
```

Several (new) functions from GEOS are now exposed in Shapely:

* :func:`.hausdorff_distance` and :func:`.frechet_distance`
* :func:`.contains_properly`
* :func:`.extract_unique_points`
* :func:`.reverse`
* :func:`.build_area` (GEOS >= 3.8)
* :func:`.minimum_bounding_circle` and :func:`.minimum_bounding_radius` (GEOS >= 3.8)
* :func:`.coverage_union` and :func:`.coverage_union_all` (GEOS >= 3.8)
* :func:`.segmentize` (GEOS >= 3.10)
* :func:`.dwithin` (GEOS >= 3.10)

In addition some aliases for existing methods have been added to provide a
method name consistent with GEOS or PostGIS:

- :func:`.line_interpolate_point` (``interpolate``)
- :func:`.line_locate_point` (``project``)
- :func:`.offset_curve` (``parallel_offset``)
- :func:`.point_on_surface` (``representative_point``)
- :func:`.oriented_envelope` (``minimum_rotated_rectangle``)
- :func:`.delaunay_triangles` (``ops.triangulate``)
- :func:`.voronoi_polygons` (``ops.voronoi_diagram``)
- :func:`.shortest_line` (``ops.nearest_points``)
- :func:`.is_valid_reason` (``validation.explain_validity``)

Getting information / parts / coordinates from geometries
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A set of GEOS getter functions are now also exposed to inspect geometries:

Reviewer: Suggest making this into a bulleted list like above

:func:`.get_dimensions`, :func:`.get_coordinate_dimension`,
:func:`.get_srid`, :func:`.get_num_points`, :func:`.get_num_interior_rings`,
:func:`.get_num_geometries`, :func:`.get_num_coordinates`,
:func:`.get_precision`.

> Several functions are added to extract parts: :func:`.get_geometry` to get a

Reviewer: Also suggest making this into a list

:func:`.get_z` to get the x/y/z coordinate of a Point.

Methods to extract all parts or coordinates at once have been added:

* The :func:`.get_parts` function can be used to get individual parts of an
  array of multi-part geometries.
* The :func:`.get_rings` function, similar as ``get_parts`` but specifically
  to extract the rings of Polygon geometries.
* The :func:`.get_coordinates` function to get all coordinates from a
  geometry or array of goemetries as an array of floats.

Each of those three functions has an optional ``return_index`` keyword, which
allows to also return the indexes of the original geometries in the source
array.

Prepared geometries
~~~~~~~~~~~~~~~~~~~

Prepared geometries are now no longer separate objects, but geometry objects
themselves can be prepared (this makes the ``shapely.prepared`` module
superfluous).

The :func:`.prepare()` function generates a GEOS prepared geometry which is
stored on the Geometry object itself. All binary predicates (except
``equals``) will make use of this if the input geometry has already been
prepared. Helper functions :func:`.destroy_prepared` and :func:`.is_prepared`
are also available.

GeoJSON IO
~~~~~~~~~~

> * Added GeoJSON input/output capabilities (:func:`.from_geojson`,

Reviewer: ```suggestion
* Added GeoJSON input/output capabilities :func:`.from_geojson` and
```

(no need for parens)

>   :func:`.to_geojson`) for GEOS >= 3.10 (#413).

Reviewer: ```suggestion
  :func:`.to_geojson` for GEOS >= 3.10 (#413).
```
